File: LeetCode/Medium/82 - Remove Duplicates from Sorted List II.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
    let cur = head, pre = null;
    while (cur && cur.next) {
        if (cur.val == cur.next.val) {
            while (cur && cur.next && cur.val == cur.next.val) cur = cur.next;
            cur = cur.next;
            if (!pre) head = cur;
            else pre.next = cur;
        }
        else {
            pre = cur;
            cur = cur.next;    
        }
    }
    return head;
};

File: LeetCode/Medium/82 - Remove Duplicates from Sorted List II/82 - Remove Duplicates from Sorted List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode *cur = head, *pre = nullptr;
        while (cur && cur->next) {
            if (cur->val == cur->next->val) {
                while (cur && cur->next && cur->val == cur->next->val) 
                    cur = cur->next;
                cur = cur->next;
                if (!pre) 
                    head = cur;
                else 
                    pre->next = cur;
            }
            else {
                pre = cur;
                cur = cur->next;    
            }
        }
        return head;
    }
};